server/src/main/java/org/elasticsearch/common/util/BytesRefArray.java

```diff
@@ -106,6 +113,15 @@ public void close() {
         Releasables.close(bytes, startOffsets);
     }
 
+    public static BytesRefArray takeOwnershipOf(BytesRefArray other) {
+        BytesRefArray b = new BytesRefArray(other.startOffsets, other.bytes, other.size, other.bigArrays);
+        other.bytes = null;
+        other.startOffsets = null;
```

Comment:
This leaves the `other` object in a state where calling `get()`, `writeTo()` or `ramBytesUsed()` on it will throw an NPE.  Previously it could never have been in such a state.

It's probably worth adding some assertions to these methods to prove that this never happens in tests and state clearly what the mistake is if it ever does happen.  I.e. something like this in each of the affected methods:

```
     assert startOffsets != null : "using BytesRefArray after ownership taken";
```

Reply:
I am about to PR these core changes separately, this one in https://github.com/elastic/elasticsearch/pull/87660

yes, I am unsure about the behavior of this. Although this passed array should not be used afterwards, the potential NPE is odd. I therefore allocate a 1 size array here in the PR there.

May you take a look at the PR above?
Another one for the other core change (`BytesRefStreamOutput`) is in the works.